DEMO_ICASSP2021/dsf_lite_lehmer.m

```diff
@@ -0,0 +1,204 @@
+    [M, K] = size(X);
+
+    %-- Whitening
+    [X, Q_PCA, Q_IPCA] = whitening(X);
+
+    %-- Normalize X
+    column_norm_X = sqrt(sum(conj(X) .* X, 1) + o.norm_eps);
+    X = bsxfun(@rdivide, X, column_norm_X);
+
+    %-- Initialization
+    if isempty(o.A_init)
+        A_init = randn(M, N);
+
+        if ~isreal(X)
+            A_init = A_init + 1i * randn(M, N);
+        end
+
+    else
+        A_init = o.A_init;
+    end
+
+    J_khl = [];
+
+    if o.khl_init
+        % Run K-hyperlines to initialize the algorithm
+        o.A_init = A_init;
+        [~, A_init, ~, J_khl, ~] = k_hyperlines(X, N, o);
+    end
+
+    A_init = cast(A_init, class(X));
+    column_norm_A_init = sqrt(sum(conj(A_init) .* A_init, 1) + o.norm_eps);
+    A_init = bsxfun(@rdivide, A_init, column_norm_A_init);
+
+    [~, K] = size(X);
+    w_init = (K + (N - 1) * ralpha(2)) * ones(N, 1);
+
+    param_init = {A_init, w_init};
+
+    % Minimize the objective function of DSF
+    obj_func = @(param, X) cost_function(param, X, ralpha, o.norm_eps);
+    [best_param, J] = optimize(obj_func, param_init, X, o.max_epoch, o.opt_tol, ...
+        o.prog_tol, o.corrections, o.deriv_check, ...
+        o.use_mex, o.verbose);
+    best_A = best_param{1};
+    % best_w = best_param{2};
+
+    % Normalization
+    [U, S, V] = svd(best_A, 'econ');
+    best_A = U * V'; % best_A = (best_A*best_A')^-1/2*best_A;
+    column_norm_best_A = sqrt(sum(conj(best_A) .* best_A, 1) + o.norm_eps);
+    best_A = bsxfun(@rdivide, best_A, column_norm_best_A);
+
+    % Returns
+    % J = [J_khl; J];
+    What = best_A' * Q_PCA; % Valid for determined case only
+    Ahat = Q_IPCA * best_A;
+    cos2 = abs(best_A' * X).^2;
+end
+
+function [cost, grad] = cost_function(param, X, ralpha, norm_eps)
+    A = param{1}; w = param{2};
+    r = ralpha(1); alpha = ralpha(2);
+
+    [M, N] = size(A);
+    [~, K] = size(X);
+
+    Aorth = A;
+    [U, S, V] = svd(A, 'econ');
+    % row-wise decoupling
+    Aorth = U * V'; % Aorth = (A*A')^-1/2*A;
+
+    % column norm normalization
+    column_norm_A = sqrt(sum(conj(Aorth) .* Aorth, 1) + norm_eps);
+    Anorm = bsxfun(@rdivide, Aorth, column_norm_A);
+
+    % Compute loss and gradient
+    angle_ = Anorm' * X;
+    Y2 = 1 - abs(angle_).^2;
+    [Y2_min, min_idx] = min(Y2, [], 1);
+
+    w(w < 0) = 0; w = w + alpha;
+    [Y2_min, minIdx] = min(Y2, [], 1);
+
+    Y2_contra = Y2 ./ Y2_min;
+    Y2_contra(minIdx(Y2_min == 0), Y2_min == 0) = 1;
+
+    Dr2 = Y2_contra.^(r - 2);
+    Dr1 = Dr2 .* Y2_contra;
+    Dr = Dr1 .* Y2_contra;
+
+    wSumDr = sum(w .* Dr, 1);
+    wSumDr1 = sum(w .* Dr1, 1);
+
+    g = wSumDr ./ wSumDr1;
+    g(Y2_min == 0) = 0;
+    g = sum(Y2_min .* g, 2);
+
+    dgdD = (w .* Dr2 ./ wSumDr1.^2) .* ((r - 1) .* wSumDr - r .* Y2_contra .* wSumDr1);
+    dgdD(:, Y2_min == 0) = 0;
+
+    dgdw = Y2_min .* (Dr .* wSumDr1 - Dr1 .* wSumDr) ./ (wSumDr1.^2) ./ (K + alpha * N);
+    dgdw(:, Y2_min == 0) = 0;
+    gradw = mean(dgdw, 2);
+
```

Comment:
condition if w < 0

Reply:
it's already conditioned all the way up above in the code. (line 166)
```
w(w < 0) = 0; w = w + alpha;
```